tensorflow/contrib/go/tensor.go

```diff
@@ -0,0 +1,753 @@
+	default:
+		return nil, &ErrTensorTypeNotSupported{
+			TensotType: dataType,
+		}
+	}
+
+	tensor, err := newTensor(dataType, ts, unsafe.Pointer(dataPtr), int64(len(dataSer))*dataSize)
+	if err != nil {
+		return nil, err
+	}
+
+	tensor.FloatVal = auxTensor.FloatVal
+	tensor.DoubleVal = auxTensor.DoubleVal
+	tensor.IntVal = auxTensor.IntVal
+	tensor.StringVal = auxTensor.StringVal
+	tensor.ScomplexVal = auxTensor.ScomplexVal
+	tensor.Int64Val = auxTensor.Int64Val
+	tensor.BoolVal = auxTensor.BoolVal
+
+	return tensor, nil
+}
+
+// DataType returns the data type of the elements contained in the tensor.
+func (t *Tensor) DataType() DataType {
+	return DataType(TF_TensorType(t.tensor))
+}
+
+// NumDims returns the number of dimensions in tensor t.
+func (t *Tensor) NumDims() int {
+	return TF_NumDims(t.tensor)
+}
+
+// Shape returns the shape of the tensor.
+func (t *Tensor) Shape() TensorShape {
+	if t.NumDims() == 0 {
+		// This is a scalar tensor
+		return [][]int64{}
+	}
+
+	shape := make([][]int64, t.NumDims())
+	for i := 0; i < t.NumDims(); i++ {
+		shape[i] = []int64{int64(t.Dim(i))}
+	}
+
+	return shape
+}
+
+// Dim returns the size of the specified dimension.
+func (t *Tensor) Dim(n int) int {
+	return int(TF_Dim(t.tensor, n))
+}
+
+// DataSize returns the size of the data in bytes contained in a tensor.
+func (t *Tensor) DataSize() int64 {
+	return TF_TensorByteSize(t.tensor)
+}
+
+// Data returns the data contained in a tensor as bytes slice.
+func (t *Tensor) Data() []byte {
+	length := t.DataSize()
+	return (*[1 << 40]byte)(unsafe.Pointer(TF_TensorData(t.tensor)))[:length:length]
+}
+
+// String returns a human-readable string description of a Tensor.
+func (t *Tensor) String() string {
+	return fmt.Sprintf("%v: dims:%v size:%v", t.DataType(), t.NumDims(), t.DataSize())
+}
+
+// ByteSlices returns the Tensor content as strings slice if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
+// The datatypes are:
+//  - DTString
+func (t *Tensor) ByteSlices() ([][]byte, error) {
+	if DTString != t.DataType() {
+		return nil, &ErrInvalidTensorType{
+			TensorType:   t.DataType(),
+			ExpectedType: DTString,
+		}
+	}
+
+	if t.StringVal != nil {
+		return t.StringVal, nil
+	}
+
+	resultBytes := []byte{}
+	inStr := false
+	t.StringVal = [][]byte{}
+	for _, b := range t.Data() {
+		if inStr {
+			if b == cBellByte {
+				t.StringVal = append(t.StringVal, resultBytes)
+				resultBytes = []byte{}
+			} else {
+				resultBytes = append(resultBytes, byte(b))
+			}
+		} else {
+			// TODO: Must be any better way to parse the strings...
+			if b == cAckByte || b == cBellByte || b == cDc1 {
+				inStr = true
+			}
+		}
+	}
+	if len(resultBytes) > 0 {
+		t.StringVal = append(t.StringVal, resultBytes)
+	}
+	t.Dtype = pb.DataType(TF_TensorType(t.tensor))
+
+	return t.StringVal, nil
+}
+
+// Float32s returns the Tensor content as float32 slice if the tensor
+// type matches, if not returns a ErrInvalidTensorType error.
```

Comment:
Same thing - probably apply throughout here.  Instead of the generic "if the tensor type matches", if the tensor type is DTFloat, etc.